### ISSUE TITLE
Add Back Propagation

### ISSUE THREAD

**Issue by StickOnAStick** [StickOnAStick/BranchPrediction#11]
If we want to learn we must train.

Back propagation in transformers is a black box for the all of us, but is critical for a transformer to function.

Back-Prop abstractly consists of three primary steps:

1.) Hold Intermediate States (In/Out)
2.) Compute loss function via BackWards pass (likely log-likelyhood or gradient descent)
3.) Update weights and save to file

# Concerns
## Primary
We can update the weights file all we want but the transformer **_instantiation_** will still be using the originally loaded weights. 

How can we ""refresh"" the transformer's parameters `w_ff1, w_ff2, w_q ...etc`

**Comment by StickOnAStick** [comment]
Concern is mostly already remediated since we store all of the weight matrices within `TransFormer::`
